primitives/src/rescue/sponge.rs

```diff
@@ -0,0 +1,420 @@
+//! This file contains the APIs wrappers for ark-sponge
+
+use ark_ff::{BigInteger, PrimeField};
+use ark_sponge::{
+    Absorb, CryptographicSponge, FieldBasedCryptographicSponge, FieldElementSize, SpongeExt,
+};
+use ark_std::{string::ToString, vec, vec::Vec};
+use jf_utils::{field_switching, pad_with_zeros};
+use num_bigint::BigUint;
+
+use super::{errors::RescueError, Permutation, RescueParameter, RescueVector, RATE, STATE_SIZE};
+
+#[derive(Clone, Default)]
+/// A rescue hash function consists of a permutation function and
+/// an internal state.
+pub struct RescueSponge<F: RescueParameter, const ABSORB_RATE: usize> {
+    pub(crate) state: RescueVector<F>,
+}
+
+/// CRHF
+pub type RescueSpongeCRHF<F> = RescueSponge<F, 3>;
+/// PRF
+pub type RescueSpongePRF<F> = RescueSponge<F, 4>;
+
+impl<F: RescueParameter, const CHUNK_SIZE: usize> RescueSponge<F, CHUNK_SIZE> {
```

Comment:
This `CHUNK_SIZE` seems not used in the following `impl`.
Also previously it's called `ABSORB_RATE` in the struct definition. Perhaps it's better to have a unified name? E.g. we could call it `RATE`, and rename `super::RATE` to `super::RESCUE_CRHF_RATE`.

Reply:
I unified it, but to `CHUNK_SIZE`. This is an internal detail of what the absorb phase does, it splits things into chunks. But I understand the confusion and it's a very valid point.
This prompted me to go and enforce that `SpongeCRH` is **always** using a `CHUNK_SIZE = 3`, while the PRF is using `CHUNK_SIZE = 4`: see https://github.com/EspressoSystems/jellyfish/pull/146/commits/ef81c461e154bb3ba8041b7406d8a9d03ca1d954 & https://github.com/EspressoSystems/jellyfish/pull/146/commits/97e9c0899351196b7e8292bd10125bcf183373bb.
Let me know what you think.

Reply:
I also personally find `ABSORB_RATE` a clearer name.

either way, we need to add code documentation somewhere (non-public, just `//`) to explain the differences between these parameterizations.

Reply:
@mrain After some thought, I went with your suggestion and renamed it to `RATE`.
@alxiong Since the sponge actually uses this for both absorb and squeeze step, I think `RATE` is more appropriate than `ABSORB_RATE` in the end. The reason it was `ABSORB_RATE` in the initial version was because of a mismatch between our usage (always squeeze 1 element) and the general squeeze API, which was pointed out by @chancharles92 [in this comment](https://github.com/EspressoSystems/jellyfish/pull/146#discussion_r1035996035).